### PR TITLE
feat: add watch support that tracks tests

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -372,45 +376,51 @@ export class Extension {
...
     // So we collect all requests and sort them out in the microtask.
-    if (!this._projectsScheduledToRun) {
-      this._projectsScheduledToRun = [];
-      this._projectsScheduledToRun.push(project);
+    if (!this._itemsScheduledToRun) {
+      this._itemsScheduledToRun = [];
+      this._itemsScheduledToRun.push({ project, request });
       // Make sure to run tests outside of this function's control flow
       // so that we can create a new TestRunRequest and see its output.
       // TODO: remove once this is fixed in VSCode (1.78?) and we
```
Comment (mxschmitt) [inline src/extension.ts:392]: if we bump, we can also remove this hack I guess?

File: src/watchSupport.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        for (const watch of this._watches) {
+          for (const i of watch.request.include || []) {
+            if (isAncestorOf(include, i)) {
+              this._watches.delete(watch);
+              break;
+            }
+          }
+        }
+      }
+      console.log(this._watches);
```
Comment (mxschmitt) [inline src/watchSupport.ts:56]: debug?

File: src/extension.ts [hunk 3]
```diff
@@ -372,45 +376,51 @@ export class Extension {
...
+    const projects = testRunInfos.map(info => info.project);
+    const include = testRunInfos.map(info => info.request.include || []).flat();
+
     // Create a test run that potentially includes all the test items.
     // This allows running setup tests that are outside of the scope of the
     // selected test items.
     const rootItems: vscodeTypes.TestItem[] = [];
     this._testController.items.forEach(item => rootItems.push(item));
-    const requestWithDeps = new this._vscode.TestRunRequest(rootItems, [], request.profile);
+    const requestWithDeps = new this._vscode.TestRunRequest(rootItems, [], undefined, mode === 'watch');
```
Comment (dgozman) [inline src/extension.ts:418]: What does `mode === 'watch'` mean here?

File: src/extension.ts [hunk 4]
```diff
@@ -520,7 +530,12 @@ located next to Run / Debug Tests toolbar buttons.`);
       await model.workspaceChanged(change);
     // Workspace change can be deferred, make sure editors are
     // decorated.
-    this._updateVisibleEditorItems();
+    await this._updateVisibleEditorItems();
+    this._watchSupport.workspaceChanged(change);
+  }
+
+  private _watchesTriggered(watches: Watch[]) {
+    this._runMatchingTests(watches, 'watch');
```
Comment (dgozman) [inline src/extension.ts:538]: We probably should enqueue watch runs.

**Review by dgozman — APPROVED** [review]

File: src/extension.ts (hunk 3 above)
Comment (pavelfeldman) [reply src/extension.ts:418]: 4th parameter is continuous.

File: package.json [hunk 5]
```diff
@@ -119,7 +119,7 @@
     "@types/glob": "^8.0.0",
     "@types/node": "^18.11.9",
     "@types/stack-utils": "^2.0.1",
-    "@types/vscode": "1.73.0",
+    "@types/vscode": "1.85.0",
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply package.json:122]: follow up?

File: src/extension.ts (hunk 1 above)
Comment (pavelfeldman) [reply src/extension.ts:392]: follow up?

File: src/extension.ts (hunk 4 above)
Comment (pavelfeldman) [reply src/extension.ts:538]: I'll queue them in a promise chain for now.

File: src/playwrightTest.ts [hunk 6]
```diff
@@ -170,8 +170,12 @@ export class PlaywrightTest {
     return { entries, errors };
   }
 
+  private _useTestServer(config: TestConfig) {
+    return config.version >= 1.43 || this._settingsModel.useTestServer.get();
```
Comment (dgozman) [inline src/playwrightTest.ts:174]:
```suggestion
    return config.version >= 1.43 && this._settingsModel.useTestServer.get();
```

**Review by dgozman — APPROVED** [review]

**Pull request merged**